File: src/main/kotlin/dev/murad/shipping/util/InventoryUtils.java
package dev.murad.shipping.util;

import dev.murad.shipping.setup.ModDataComponents;
import net.minecraft.world.Container;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.energy.IEnergyStorage;
import net.neoforged.neoforge.items.ItemStackHandler;
import org.checkerframework.checker.nullness.qual.NonNull;


import javax.annotation.Nullable;
import java.util.*;

public class InventoryUtils {

    public static boolean mayMoveIntoInventory(Container target, Container source) {
        if (source.isEmpty()) {
            return false;
        }

        HashMap<Item, List<ItemStack>> map = new HashMap<>();
        List<Integer> airList = new ArrayList<>();
        for (int i = 0; i < target.getContainerSize(); i++) {
            ItemStack stack = target.getItem(i);
            if ((stack.isEmpty() || stack.getItem().equals(Items.AIR)) && target.canPlaceItem(i, stack)) {
                airList.add(i);
            } else if (stack.getMaxStackSize() != stack.getCount() && target.canPlaceItem(i, stack)) {
                if (map.containsKey(stack.getItem())) {
                    map.get(stack.getItem()).add(stack);
                } else {
                    map.put(stack.getItem(), new ArrayList<>(Collections.singleton(stack)));
                }
            }
        }

        for (int i = 0; i < source.getContainerSize(); i++) {
            ItemStack stack = source.getItem(i);
            if (!stack.isEmpty() && map.containsKey(stack.getItem())) {
                for (ItemStack targetStack : map.get(stack.getItem())) {
                    if (canMergeItems(targetStack, stack))
                        return true;
                }
            } else if (!airList.isEmpty() && target instanceof Entity) {
                Entity e = (Entity) target;
                boolean validSlot = Optional.ofNullable(e.getCapability(Capabilities.ItemHandler.ENTITY))
                        .map(itemHandler -> airList.stream()
                                .map(j -> itemHandler.isItemValid(j, stack))
                                .reduce(false, Boolean::logicalOr)).orElse(true);
                if (validSlot) {
                    return true;
                }
            } else if (!airList.isEmpty()) {
                return true;
            }
        }
        return false;
    }

    public static boolean isEmpty(ItemStackHandler itemHandler) {
        for (int i = 0; i < itemHandler.getSlots(); i++) {
            if (!itemHandler.getStackInSlot(i).isEmpty() && !itemHandler.getStackInSlot(i).getItem().equals(Items.AIR)) {
                return false;
            }
        }
        return true;
    }

    @NonNull
    public static ItemStack moveItemStackIntoHandler(ItemStackHandler handler, @NonNull ItemStack stack) {
        var slots = handler.getSlots();
        for (int i = 0; i < slots && !stack.isEmpty(); i++) {
            stack = handler.insertItem(i, stack, false);
        }
        return stack;
    }

    public static boolean canMergeItems(ItemStack stack1, ItemStack stack2) {
        if (stack1.getItem() != stack2.getItem()) {
            return false;
        } else if (stack1.getDamageValue() != stack2.getDamageValue()) {
            return false;
        } else if (stack1.getCount() > stack1.getMaxStackSize()) {
            return false;
        } else {
            return ItemStack.isSameItemSameComponents(stack1, stack2);
        }
    }

    @Nullable
    public static IEnergyStorage getEnergyCapabilityInSlot(int slot, ItemStackHandler handler) {
        ItemStack stack = handler.getStackInSlot(slot);
        if (!stack.isEmpty()) {
            Optional<IEnergyStorage> energyStorage = Optional.ofNullable(
                    stack.getComponents().get(ModDataComponents.ENERGY.get())
            );
            if (energyStorage.isPresent()) {
                return energyStorage.get();
            }
        }
        return null;
    }
}

File: src/main/kotlin/dev/murad/shipping/util/InventoryUtils.kt
package dev.murad.shipping.util

import dev.murad.shipping.setup.ModDataComponents
import net.minecraft.world.Container
import net.minecraft.world.entity.Entity
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.neoforged.neoforge.capabilities.Capabilities
import net.neoforged.neoforge.energy.IEnergyStorage
import net.neoforged.neoforge.items.IItemHandler
import net.neoforged.neoforge.items.ItemStackHandler
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.MutableList
import kotlin.collections.set
import kotlin.collections.setOf

object InventoryUtils {

    fun mayMoveIntoInventory(target: Container, source: Container): Boolean {
        if (source.isEmpty) {
            return false
        }

        val map = HashMap<Item, MutableList<ItemStack>>()
        val airList: MutableList<Int> = ArrayList()
        for (i in 0 until target.containerSize) {
            val stack = target.getItem(i)
            if ((stack.isEmpty || stack.item == Items.AIR) && target.canPlaceItem(i, stack)) {
                airList.add(i)
            } else if (stack.maxStackSize != stack.count && target.canPlaceItem(i, stack)) {
                if (map.containsKey(stack.item)) {
                    map[stack.item]!!.add(stack)
                } else {
                    map[stack.item] = ArrayList(setOf(stack))
                }
            }
        }

        for (i in 0 until source.containerSize) {
            val stack = source.getItem(i)
            if (!stack.isEmpty && map.containsKey(stack.item)) {
                for (targetStack in map[stack.item]!!) {
                    if (canMergeItems(targetStack, stack)) return true
                }
            } else if (!airList.isEmpty() && target is Entity) {
                val e = target as Entity
                val validSlot = Optional.ofNullable(e.getCapability(Capabilities.ItemHandler.ENTITY))
                    .map { itemHandler: IItemHandler ->
                        airList.stream()
                            .map { j: Int? -> itemHandler.isItemValid(j!!, stack) }
                            .reduce(false) { a: Boolean, b: Boolean -> java.lang.Boolean.logicalOr(a, b) }
                    }.orElse(true)
                if (validSlot) {
                    return true
                }
            } else if (!airList.isEmpty()) {
                return true
            }
        }
        return false
    }

    fun isEmpty(itemHandler: ItemStackHandler): Boolean {
        for (i in 0 until itemHandler.slots) {
            if (!itemHandler.getStackInSlot(i).isEmpty && itemHandler.getStackInSlot(i).item != Items.AIR) {
                return false
            }
        }
        return true
    }

    fun moveItemStackIntoHandler(handler: ItemStackHandler, stack: ItemStack): ItemStack {
        var stack = stack
        val slots = handler.slots
        var i = 0
        while (i < slots && !stack.isEmpty) {
            stack = handler.insertItem(i, stack, false)
            i++
        }
        return stack
    }

    fun canMergeItems(stack1: ItemStack, stack2: ItemStack): Boolean {
        return if (stack1.item !== stack2.item) {
            false
        } else if (stack1.damageValue != stack2.damageValue) {
            false
        } else if (stack1.count > stack1.maxStackSize) {
            false
        } else {
            ItemStack.isSameItemSameComponents(stack1, stack2)
        }
    }

    fun getEnergyCapabilityInSlot(slot: Int, handler: ItemStackHandler): IEnergyStorage? {
        val stack = handler.getStackInSlot(slot)
        if (!stack.isEmpty) {
            val energyStorage = Optional.ofNullable<IEnergyStorage>(
                stack.components.get(ModDataComponents.getEnergyStorage().get())
            )
            if (energyStorage.isPresent) {
                return energyStorage.get()
            }
        }
        return null
    }
}
